### DISCUSSION TITLE
Embedding xsd schemas

### DISCUSSION THREAD

**Discussion by andrew-hardin** [ngageoint/six-library#543] [Ideas]
There's a fair amount of brittle book keeping required to track the location of schemas and resolve them at run-time. An alternative approach involves embedding the schemas into the source (e.g. `xxd -i schema.xsd`), then at run-time passing those buffers to the xml validator.

Is that a feature that may be valuable to the broader community? ~If so, is it a feature that replaces the existing schema path machinery, or, does that path based machinery enable a valuable use case that I'm ignorant to?~ If so, it probably makes sense to maintain the existing schema paths for backwards compatibility.

**Comment by JDanielSmith** [comment]
Yes, the schema-path machinery is a bit complicated; I like the idea of embedding a schema! But this probably needs to get run through "the process" (although I'm not sure exactly what that might be) as an enhancement.

This probably isn't possible in a standard way unless/until [`std::embed`/`#embed`](http://open-std.org/JTC1/SC22/WG21/docs/papers/2020/p1040r6.html) is accepted.

**Reply by andrew-hardin** [reply]
I see where you're going with `std::embed`. That certainly would be a clean way to bake a resource into the binary. However, using CMake's `configure_file` could also get you to approximately the same destination -- using the build system instead of a built-in compiler feature. I'll try to mock up a proof-of-concept to demo the idea.

**Reply by JDanielSmith** [reply]
I think we still need to support `waf` builds :-(, so a CMake-only solution probably won't work.

**Reply by andrew-hardin** [reply]
The `waf` requirement is unfortunate. The CMake-only option isn't too bad, and the resulting `.h/.cpp` are surprisingly clean: https://github.com/andrew-hardin/six-library/commit/e3f396aafa515ba89cf740c85ad99ad4ffe40c82
